Mathlib/NumberTheory/ModularForms/Petersson.lean

```diff
@@ -0,0 +1,82 @@
+/-
+Copyright (c) 2025 David Loeffler. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: David Loeffler
+-/
+
+import Mathlib.NumberTheory.ModularForms.Basic
+
+/-!
+# The Petersson scalar product
+
+For `f, f'` functions `ℍ → ℂ`, we define `petersson k f f'` to be the function
+`τ ↦ conj (f τ) * f' τ * τ.im ^ k`.
+
+We show this function is (weight 0) invariant under `Γ` if `f, f'` are (weight `k`) invariant under
+`Γ`.
+-/
+
+
+open UpperHalfPlane
+
+open scoped MatrixGroups ComplexConjugate ModularForm
+
+namespace UpperHalfPlane
+
+/-- The integrand in the Petersson scalar product of two modular forms. -/
+noncomputable def petersson (k : ℤ) (f f' : ℍ → ℂ) (τ : ℍ) :=
+  conj (f τ) * f' τ * τ.im ^ k
+
+lemma petersson_continuous (k : ℤ) {f f' : ℍ → ℂ} (hf : Continuous f) (hf' : Continuous f') :
+    Continuous (petersson k f f') := by
+  apply ((Complex.continuous_conj.comp hf).mul hf').mul
+  apply (Complex.continuous_ofReal.comp continuous_im).zpow₀
+  exact fun τ ↦ .inl <| Complex.ofReal_ne_zero.mpr τ.im_ne_zero
```

Comment:
```suggestion
attribute [fun_prop] Complex.continuous_conj UpperHalfPlane.continuous_im

lemma petersson_continuous (k : ℤ) {f f' : ℍ → ℂ} (hf : Continuous f) (hf' : Continuous f') :
    Continuous (petersson k f f') := by
  unfold petersson
  have (τ : ℍ) : _ ∨ 0 ≤ k := Or.inl <| Complex.ofReal_ne_zero.mpr τ.im_ne_zero
  fun_prop (disch := assumption)
```
The `fun_prop` attributes should be added directly at the definitions, of course.